File: Core/src/main/java/com/hqq/core/recycler/adapter/RecycleBannerAdapter.java
package com.hqq.core.recycler.adapter;

import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.hqq.core.R;
import com.hqq.core.glide.ImageLoadUtils;
import com.hqq.core.recycler.RecycleViewBanner;

import java.util.ArrayList;
import java.util.List;

/**
 * @Author : huangqiqiang
 * @Package : com.core.library.banner
 * @FileName :   BaseBannerAdapter
 * @Date : 2018/6/15 0015  下午 5:26
 * @Descrive : TODO
 * @Email :  qiqiang213@gmail.com
 */
public class RecycleBannerAdapter<T> extends RecyclerView.Adapter {
    private List<T> mData = new ArrayList<>();
    private RecycleViewBanner.RecycleViewBannerClickListener onRvBannerClickListener;
    private RecycleViewBanner.RecycleViewBannerChangeListener mRecycleViewBannerChangeListener;
    private boolean mIsShowTip;
    private boolean mIsUnlimited = true;

    public void setUnlimited(boolean unlimited) {
        mIsUnlimited = unlimited;
    }

    public void setData(List<T> data) {
        mData = data;
    }

    public void setOnRvBannerClickListener(RecycleViewBanner.RecycleViewBannerClickListener onRvBannerClickListener) {
        this.onRvBannerClickListener = onRvBannerClickListener;
    }

    public RecycleBannerAdapter setRecycleViewBannerChangeListener(RecycleViewBanner.RecycleViewBannerChangeListener recycleViewBannerChangeListener) {
        mRecycleViewBannerChangeListener = recycleViewBannerChangeListener;
        return this;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = onCreateView(parent);
        return new RecyclerView.ViewHolder(view) {
        };
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {

        onBindView(holder, position);

    }


    @Override
    public int getItemCount() {
        if (mIsUnlimited) {
            return mData == null ? 0 : mData.size() < 2 ? mData.size() : Integer.MAX_VALUE;
        } else {
            return mData == null ? 0 : mData.size();
        }
    }

    public void setIsShowTip(boolean isShowTip) {
        mIsShowTip = isShowTip;
    }

    public boolean isShowTip() {
        return mIsShowTip;
    }

    public void setShowTip(boolean isShowTip) {
        mIsShowTip = isShowTip;
    }


    private View onCreateView(ViewGroup parent) {
        return LayoutInflater.from(parent.getContext()).inflate(R.layout.item_banner, parent, false);
    }

    private void onBindView(RecyclerView.ViewHolder holder, int position) {
        final int finalPosition = position;
        ImageView img = holder.itemView.findViewById(R.id.iv_banner);
        TextView tv = holder.itemView.findViewById(R.id.tv_code_banner);
        T data = mData.get(position % mData.size());
        if (data instanceof String) {
            tv.setVisibility(View.GONE);
            ImageLoadUtils.with((String) data, img);
        } else if (data instanceof Integer) {
            tv.setVisibility(View.GONE);
            ImageLoadUtils.with((Integer) data, img);
        } else if (null != mRecycleViewBannerChangeListener) {
            ImageLoadUtils.with(mRecycleViewBannerChangeListener.getUrl(data), img);
            if (mIsShowTip) {
                tv.setText(mRecycleViewBannerChangeListener.getTitle(data));
                tv.setVisibility(View.VISIBLE);
            } else {
                tv.setVisibility(View.GONE);
            }
        }

        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (onRvBannerClickListener != null) {
                    onRvBannerClickListener.onBannerClick(finalPosition % mData.size());
                }
            }
        });
    }

}

File: Core/src/main/java/com/hqq/core/recycle/adapter/RecycleBannerAdapter.kt
package com.hqq.core.recycle.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.hqq.core.R
import com.hqq.core.glide.ImageLoadUtils
import com.hqq.core.recycle.RecycleViewBanner.RecycleViewBannerChangeListener
import com.hqq.core.recycle.RecycleViewBanner.RecycleViewBannerClickListener

/**
 * @Author : huangqiqiang
 * @Package : com.core.library.banner
 * @FileName :   BaseBannerAdapter
 * @Date : 2018/6/15 0015  下午 5:26
 * @Descrive : TODO
 * @Email :  qiqiang213@gmail.com
 */
class RecycleBannerAdapter<Any> : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    lateinit var mData: MutableList<Any>
    private var onRvBannerClickListener: RecycleViewBannerClickListener<*>? = null
    private var mRecycleViewBannerChangeListener: RecycleViewBannerChangeListener<Any>? = null
    var isShowTip = false

    private var mIsUnlimited = true
    fun setUnlimited(unlimited: Boolean) {
        mIsUnlimited = unlimited
    }

    fun setData(data: MutableList<Any>?) {
        data?.let {
            mData = it
        }
    }

    fun setOnRvBannerClickListener(onRvBannerClickListener: RecycleViewBannerClickListener<*>) {
        this.onRvBannerClickListener = onRvBannerClickListener
    }

    fun setRecycleViewBannerChangeListener(recycleViewBannerChangeListener: RecycleViewBannerChangeListener<Any>?): RecycleBannerAdapter<Any> {
        mRecycleViewBannerChangeListener = recycleViewBannerChangeListener
        return this
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val view = onCreateView(parent)
        return object : RecyclerView.ViewHolder(view) {}
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        onBindView(holder, position)
    }

    override fun getItemCount(): Int {
        return if (mIsUnlimited) {
            if (mData == null) 0 else if (mData.size < 2) mData.size else Int.MAX_VALUE
        } else {
            if (mData == null) 0 else mData.size
        }
    }


    private fun onCreateView(parent: ViewGroup): View {
        return LayoutInflater.from(parent.context).inflate(R.layout.item_banner, parent, false)
    }

    private fun onBindView(holder: RecyclerView.ViewHolder?, position: Int) {
        val img = holder!!.itemView.findViewById<ImageView>(R.id.iv_banner)
        val tv = holder.itemView.findViewById<TextView>(R.id.tv_code_banner)
        val item = mData!![position % mData!!.size]
        if (item is String) {
            tv.visibility = View.GONE
            ImageLoadUtils.with(item as String, img)
        } else if (item is Int) {
            tv.visibility = View.GONE
            ImageLoadUtils.with(item as Int, img)
        } else if (null != mRecycleViewBannerChangeListener) {
            ImageLoadUtils.with(mRecycleViewBannerChangeListener!!.getUrl(item!!), img)
            if (isShowTip) {
                tv.text = mRecycleViewBannerChangeListener!!.getTitle(item!!)
                tv.visibility = View.VISIBLE
            } else {
                tv.visibility = View.GONE
            }
        }
        holder.itemView.setOnClickListener {
            if (onRvBannerClickListener != null) {
                onRvBannerClickListener!!.onBannerClick(position % (mData!!.size))
            }
        }
    }


}